File: src/services/mockSmdp.ts
/**
 * Mock SM-DP+ Service
 *
 * In-process module that returns canned eSIM activation codes
 * and profile metadata. Not a separate HTTP service.
 */

export interface ActivationCodeResponse {
  activationCode: string;
  iccid: string;
  smdpAddress: string;
  profileType: string;
  matchingId: string;
}

const PROFILES: Record<string, ActivationCodeResponse> = {
  default: {
    activationCode: '1$smdp.operator.com$POSTPAID-001',
    iccid: '8901010000000000001',
    smdpAddress: 'smdp.operator.com',
    profileType: 'postpaid',
    matchingId: 'POSTPAID-001',
  },
  prepaid: {
    activationCode: '1$smdp.operator.com$PREPAID-001',
    iccid: '8901010000000000002',
    smdpAddress: 'smdp.operator.com',
    profileType: 'prepaid',
    matchingId: 'PREPAID-001',
  },
  companion: {
    activationCode: '1$smdp.operator.com$COMPANION-001',
    iccid: '8901010000000000003',
    smdpAddress: 'smdp.operator.com',
    profileType: 'companion',
    matchingId: 'COMPANION-001',
  },
};

/**
 * Get an activation code for a profile key.
 * Falls back to the 'default' profile if the key is unknown.
 */
export function getActivationCode(profileKey: string): ActivationCodeResponse {
  return PROFILES[profileKey] ?? PROFILES['default']!;
}

/**
 * List all available canned profiles.
 */
export function listAvailableProfiles(): ActivationCodeResponse[] {
  return Object.values(PROFILES);
}

File: internal/services/mock_smdp.go
package services

// ActivationCodeResponse holds a canned eSIM activation code.
type ActivationCodeResponse struct {
	ActivationCode string `json:"activationCode"`
	ICCID          string `json:"iccid"`
	SMDPAddress    string `json:"smdpAddress"`
	ProfileType    string `json:"profileType"`
	MatchingID     string `json:"matchingId"`
}

var _profiles = map[string]ActivationCodeResponse{
	"default": {
		ActivationCode: "1$smdp.operator.com$POSTPAID-001",
		ICCID:          "8901010000000000001",
		SMDPAddress:    "smdp.operator.com",
		ProfileType:    "postpaid",
		MatchingID:     "POSTPAID-001",
	},
	"prepaid": {
		ActivationCode: "1$smdp.operator.com$PREPAID-001",
		ICCID:          "8901010000000000002",
		SMDPAddress:    "smdp.operator.com",
		ProfileType:    "prepaid",
		MatchingID:     "PREPAID-001",
	},
	"companion": {
		ActivationCode: "1$smdp.operator.com$COMPANION-001",
		ICCID:          "8901010000000000003",
		SMDPAddress:    "smdp.operator.com",
		ProfileType:    "companion",
		MatchingID:     "COMPANION-001",
	},
}

// GetActivationCode returns an activation code for a profile key.
// Falls back to the "default" profile if the key is unknown.
func GetActivationCode(profileKey string) ActivationCodeResponse {
	if p, ok := _profiles[profileKey]; ok {
		return p
	}
	return _profiles["default"]
}

// ListAvailableProfiles returns all canned profiles.
func ListAvailableProfiles() []ActivationCodeResponse {
	result := make([]ActivationCodeResponse, 0, len(_profiles))
	for _, p := range _profiles {
		result = append(result, p)
	}
	return result
}
